WIP on writer low level retries

ShaneSaww opened the pull request:
This is still incomplete but I wanted to push up what I had and get some review going. 
This is adding in retries to the writer that get applied in both async and sync cases. 

There is also a config value `Retries` to control it. 

TODO:
Add value to control backoff time
Docs
??

Commit:
WIP on writer low level retries

abuchanan-nr commented on writer.go:
```diff
@@ -704,8 +712,29 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
+			default:
+				attempts++
+				err = fmt.Errorf("error writing messages to %s (partition %d): %s", w.topic, w.partition, err)
+				backoff(attempts, 5*time.Millisecond, 100*time.Millisecond)
```
They have constants `backoffDelayMin` and `backoffDelayMax`

colesnodgrass replied:
Should they be configurable?

abuchanan-nr replied:
Would we change them? I kinda doubt we'd ever need different values.

ShaneSaww replied:
Yeah, maybe we should add some consts for this. 
But I don't want to delay an entire second, I want a bit faster retries

ShaneSaww replied:
yeah, there is a TODO to make they configurable as well

abuchanan-nr commented on writer.go:
```diff
@@ -704,8 +712,29 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
+
+			default:
+				attempts++
+				err = fmt.Errorf("error writing messages to %s (partition %d): %s", w.topic, w.partition, err)
```
We could log the fact that we're retrying, to the error logger.

ShaneSaww replied:
hmm, maybe retries to the normal logger but actually errors to the error logger.
I think I should also I add a stat for the amount of attempts

abuchanan-nr reviewed (commented):
Looks good.

colesnodgrass commented on writer.go:
```diff
@@ -704,8 +712,29 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
+			w.stats.errors.observe(1)
+			switch err {
+			case DuplicateSequenceNumber:
+				// we've moved beyond the batch but havn't noticed it. Just return sucess.
```
typo in `success` and `haven't`

colesnodgrass commented on writer.go:
```diff
@@ -704,8 +712,29 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
-	if _, err = conn.WriteCompressedMessages(w.codec, batch...); err != nil {
-		w.stats.errors.observe(1)
+	attempts := 0
+	for w.retries > attempts {
```
If `w.retries` is `0` you wont enter this for loop, shouldn't this be `for w.retries >= attempts`?  I guess this asks the question of whether we're defining retries as the number of retries or the number or total attempted requests?

ShaneSaww replied:
yeah, I see that. 
retries as the number of retires. 
Then I should start attempts at `-1` or change this logic completely.

ShaneSaww replied:
I changed the logic. commit incoming.

colesnodgrass commented on writer.go:
```diff
@@ -704,8 +712,29 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
+
+			default:
+				attempts++
+				err = fmt.Errorf("error writing messages to %s (partition %d): %s", w.topic, w.partition, err)
```
you're setting the err to `error writing ...` and then potentially setting the error _again_ to `error writing ...` on line `739`

ShaneSaww replied:
on 739 I'm printing it, but I am printing double. 
Fixing that.

Commit:
Adding docs, stats and logging

Commit:
Print the already formated error, and configure retry backoff

Commit:
Change logic for attempts vs retries.

This change ensures that a single attempt always happens by checking the
retry logic after we send the batch.

ShaneSaww commented:
@cole-snodgrass @abuchanan-nr I changed the retry logic, added docs, and made the backoff time configurable. this is RFR

colesnodgrass commented on writer.go:
```diff
@@ -73,6 +73,22 @@ type WriterConfig struct {
+	// Controls how many times a writer will resend records that
+	// have failed with a potentailly transient error. This gets
+	// applied to both async and non. Setting it to 0 will disable
+	// these low level retries but MaxAttempt could still have message retied.
```
typo  in `retried`

colesnodgrass commented on writer.go:
```diff
@@ -73,6 +73,22 @@ type WriterConfig struct {
 	// The default is to use a target batch size of 100 messages.
 	BatchSize int
 
+	// Controls how many times a writer will resend records that
```
`will resend` isn't entirely true, `will attempt to resend` is more exact

colesnodgrass commented on writer.go:
```diff
@@ -73,6 +73,22 @@ type WriterConfig struct {
+	// The default is 0.
+	Retries int
+
+	// The amount of time waiting before attempting to re-send a batch.
```
`re-send` or `resend`? you have both

ShaneSaww replied:
`resend` ideally

colesnodgrass commented on writer.go:
```diff
@@ -145,16 +161,18 @@ type WriterStats struct {
-	QueueCapacity     int64         `metric:"kafka.writer.queue.capacity"     type:"gauge"`
+	MaxAttempts          int64         `metric:"kafka.writer.attempts.max"       		type:"gauge"`
+	Retry                int64         `metric:"kafka.writer.retries.max"        		type:"gauge"`
+	RetryBackoffInterval time.Duration `metric:"kafka.writer.retrybackoff.interval"    	type:"gauge"`
```
Some naming inconsistencies here with the existing attributes.  `MaxAttempts` and `RequiredAcks` are plural but `Retry` is singular but the `metric` name is plural?  I would recommend renaming to `Retries` for consistency.

ShaneSaww replied:
It's because there already is a stat called `Retries`, that's watching the count of `Retries`. 

But I agree that the naming is off. I might name it something like `MaxRetries`, but that feels misleading (makes it similar to `MaxAttempts` tho).

colesnodgrass commented on writer.go:
```diff
@@ -704,10 +730,42 @@ func (w *writer) write(conn *Conn, batch []Message, resch [](chan<- error)) (ret
-		w.stats.errors.observe(1)
+	attempts := 0
+	for {
+		if _, err = conn.WriteCompressedMessages(w.codec, batch...); err != nil {
```
if `err == nil` then you're not breaking out of the for loop

ShaneSaww replied:
yep, fixing that.

Commit:
fixing naming issues, and breaking on a good send

Commit:
writing a test, it's a little strange tho

ShaneSaww commented:
@cole-snodgrass @abuchanan-nr Wanna give this a final lookover? 
I added a test, it relies on a strange case but It does show how the retries as handled.

abuchanan-nr commented:
It's confusing to have both `MaxAttempts` and `Retries` in the config and in the actual message writing code. What if the error is returned via `resch` and retries get handled in `Writer.WriteMessages`?

But, we could eat the technical debt for now and revisit this. There might be a better design for the connection code that retries each API call individually, instead of retrying the bulk `WriteMessages` code.

ShaneSaww commented:
@abuchanan-nr  I agree that it's confusing to have both retries. 
I think it's okay for us to leave this tech debt, but we need to come back and remove the `MaxAttempts` in the future. The big there is removing that is an breaking API change.
